### ISSUE TITLE
Agent showing different behavior in different episodes after the training is done.

### ISSUE THREAD

**Issue by rajatpaliwal** [Unity-Technologies/ml-agents#3217]
Hello all,
I am confused by the strange behavior I am seeing after completing the training. The agent controlled by trained brain is showing different behavior in different episodes but instead it should show the same behavior in each episode.

Any suggestion would be helpful.

Thanks,
Rajat

**Comment by andrewcoh** [comment]
The policies learned by PPO/SAC are typically stochastic.  Assuming that the different behaviors aren't significantly different in performance, this isn't unexpected.

**Comment by rajatpaliwal** [comment]
Hi @andrewcoh , thanks for replying.
In my case, I am training an agent to pass through a tunnel perfectly, without colliding with the walls. After training, in one episode agent is able to pass through a particular part of tunnel in a correct manner while in the next episode it is not doing the same or passing through that particular part including some collisions. That creates difference in performance and not acceptable.
I wonder if there is a way to remove this variation.

**Comment by andrewcoh** [comment]
Can you share your reward function?

Also, assuming your reward function encourages the agent to pass through the tunnel without colliding with anything, I'd recommend using PPO over SAC (SAC learns inherently more random policies) and to also try running for more timesteps to allow convergence.

**Comment by rajatpaliwal** [comment]
I am using PPO for training and max steps in my training are 600,000. The reward mechanism i am using is ::
Used SetReward() function for providing reward.

Time penalty (-1f/ agentParameters.maxStep)

Reward for agent facing towards target(cube at the end of tunnel) = 0.001f * m_FacingDot

Reward for moving toward target = 0.001* m_MovingTowardsDot

+0.8f reward for reaching to a minitarget with correct orientation.

-0.5f reward for reaching a minitarget with incorrect orientation.

-0.2f reward for colliding with tunnel boundary.

+1 reward for reaching the final target.

Here target is a cube at the end of tunnel , to provide positive reward to make the agent reach the end of tunnel . I am using minitargets as intermediate points at which smaller reward is provided to the agent for reaching those points with correct position and orientation.

**Comment by andrewcoh** [comment]
I'm suspicious of 
- +0.8f reward for reaching to a minitarget with correct orientation
-0.5f reward for reaching a minitarget with incorrect orientation
They might be introducing weird optima that can be hard to identify/understand.

Can you remove these/mini target and try running again? Try running for a lot more timesteps (you can kill the training run if it converges before it reaches the max steps and your model will be saved). When training, does the episodic reward seem to stabilize at a value near 1?

**Comment by rajatpaliwal** [comment]
This is the cumulative reward chart generated over the training period. 
![image](https://user-images.githubusercontent.com/30369775/72393343-72d30a00-36e7-11ea-8a3d-8a050d229657.png)
I am not sure what you mean by episodic reward.

**Comment by andrewcoh** [comment]
Hmm unfortunately that figure doesn't give a lot of information.  I'd suggest rerunning without the two intermediate rewards mentioned in my previous comment.

**Comment by rajatpaliwal** [comment]
Yes, I will be rerunning a training without those intermediate rewards. Here are all the charts generated during training to provide you with more information. Hope they are helpful.
![image](https://user-images.githubusercontent.com/30369775/72395535-b1b88e00-36ee-11ea-88f1-1e613eb97d1c.png)

**Comment by rajatpaliwal** [comment]
Hello @andrewcoh ,
I performed a training by removing the intermediate rewards and by increasing the number of max steps up to 800K. But the performance did not improved and some strange trends were observed during training. The training graphs obtained are as given below:
![image](https://user-images.githubusercontent.com/30369775/72763284-026d3280-3b98-11ea-848b-5d549db74a0b.png)
![image](https://user-images.githubusercontent.com/30369775/72763339-3a747580-3b98-11ea-807b-1c280ec3d8ab.png)
The reward mechanism I am using now is as given below:

- Time penalty (-1f/ agentParameters.maxStep)

- Reward for agent facing towards target(cube at the end of tunnel) = 0.001f * m_FacingDot

- Reward for moving toward target = 0.003f* m_MovingTowardsDot

- -0.2f reward for colliding with tunnel boundary.

- +1 reward for reaching the final target.

The agent moves very slowly after the training, and tends to turn around(full 180) and try to go out of tunnel. I am providing negative reward in order to prompt agent to finish episode fast, instead it moves very slowly and still get positive reward. I am not able to understand this behavior , any suggestion would be helpful.

**Comment by andrewcoh** [comment]
> Reward for agent facing towards target(cube at the end of tunnel) = 0.001f * m_FacingDot
may be enabling the agent to get positive reward for just facing the target (but not actually moving toward it).  I would remove this.  Secondly,
>  Reward for agent facing towards target(cube at the end of tunnel) = 0.001f * m_FacingDot
should be less  (more negative) than the timestep penalty so that the agent doesn't figure out how to constantly be moving towards the target but never actually reaching it.

**Comment by andrewcoh** [comment]
> Reward for agent facing towards target(cube at the end of tunnel) = 0.001f * m_FacingDot

may be enabling the agent to get positive reward for just facing the target (but not actually moving toward it). I would remove this. Secondly,

> Reward for moving toward target = 0.003f* m_MovingTowardsDot

should be less (more negative) than the timestep penalty so that the agent doesn't figure out how to constantly be moving towards the target but never actually reaching it.

**Comment by rajatpaliwal** [comment]
Hello @andrewcoh ,
I changed the reward mechanism . The reward mechanism I am using right now is::
Used SetReward() function for providing reward.
•	Time penalty (-0.01f) using AddReward function
•	-0.8f reward for  colliding with tunnel boundary
•	+1 reward for reaching the final target.
•      if distance to target >=12.0f -- reward = +0.01f
•      if distance to target <12.0f  and >= 8.0f  -- reward = +0.1f
•      if distance to target  <8.0f  and >= 4.0f  -- reward = +0.3f
•      if distance to target <4.0f  -- reward +0.6f
•      if speed <1.4f -- reward = -0.6f

Observations collected are ::
Vector Observation Space: 
•	Adding 12 rays each at 30 degree apart in the form of cone. Detecting objects the tunnel 
        boundary and target at the end.
•	Adding current position of the agent as an observation.
•	Adding the target position as an observation
•       Adding distance to the target as an observation
Note:: Environment reset when agent collides with the tunnel boundary or reach the final target cube at the end of tunnel.

This are the training result after training the agent for 2.5 M time steps.
![image](https://user-images.githubusercontent.com/30369775/73011819-e355e700-3dc9-11ea-8e80-48a712a0cd1e.png)
![image](https://user-images.githubusercontent.com/30369775/73011978-30d25400-3dca-11ea-86e2-84c8a5175a9f.png)
The training converged agent learned some behavior but it is not able to solve the problem completely.
![image](https://user-images.githubusercontent.com/30369775/73012355-f4ebbe80-3dca-11ea-8b3d-8282003d967b.png)
My environment is like the image given above. The agent starts from the start point and need to reach to the target while maintaining correct orientation. Right now it is just able to reach to the point as shown in the image even after 2.5 M steps. I don't think it is too difficult of a problem to be able to solve by ML-Agent. I am not sure where I am going wrong. Do you think curriculum learning would be able to solve it?

**Comment by andrewcoh** [comment]
What are the actions?

**Comment by rajatpaliwal** [comment]
The actions are::
Actions Provided:
Forward Axis: Two action-- forward movement and no action
upDownRotateAxis : Three actions -- Two Action for up and down rotation and no action
leftRightRotateAxis : Three actions -- Two Action for left and right rotation and no action

**Comment by rajatpaliwal** [comment]
I think it is too difficult for a single brain to learn the task of going up and also coming down. For this purpose it might be beneficial if I train with two brains. One for going up and other for coming down. Using GivenBrain function of the agent class.

**Comment by andrewcoh** [comment]
I think you should have no problem learning with one brain.  Do you end the episode if the agent crashes into the wall i.e. call done?

**Comment by rajatpaliwal** [comment]
Yes, I end the episode if the agent crashes into the wall.

**Comment by andrewcoh** [comment]
Is the agent within this range 

>  if distance to target <4.0f -- reward +0.6f

in that image? If so, it's more rewarding for the agent to never finish the episode because it can get .6 every timestep that it's in range of the goal.  In place of those proximity bonuses, you could use something like (max_distance - current_distance) / max_distance where max_distance is the maximum distance the agent can have to the target. The reason it is divided by max_distance is to normalize the reward to between 0 and 1.  Another thing to try that may be better is progress towards the goal.

**Comment by rajatpaliwal** [comment]
The total length of the tube is 14 units. The agent is maximum able to reach upto 8.5 units.  So it is never within this range.
if distance to target <4.0f -- reward +0.6f.
In the image it is able to reach upto 8 units only. So agent is getting +0.3f reward per time step. I can give it a try by normalizing the reward . Do you think increasing the number of hidden layer or by increasing the learning rate would help.

If you mean by progressing towards the goal without resetting the environment when collide with the wall, I have tried that too. In that case agent is never able to learn to move with correct orientation.

**Comment by andrewcoh** [comment]
Getting .3f per timestep is also more valuable than ending an episode by reaching the goal.  Did you try something like the difference between last step's distance to target and current distance to target?

**Comment by rajatpaliwal** [comment]
Hi @andrewcoh ,
Do you mean difference between last step's distance to target and current distance to target  as an observation ?
or adding the difference as a reward.

**Comment by rajatpaliwal** [comment]
Also, can time penalty of (-0.01f) is making the agent to end the episode early and not allowing to explore it to the end of the tunnel.

**Comment by rajatpaliwal** [comment]
So, it finally learned to move properly with correct orientation  in the tunnel and reaching to the end of tunnel. It took 7M steps for the agent to learn that behavior. I am sure this learning time can be brought down with fine tuning of hyperparameters.
